File: examples/elixir/phoenix-ecto/lib/autonoma_example/repositories/organization.ex
defmodule AutonomaExample.Repositories.Organization do
  # ==========================================================================
  # Organization Repository
  # ==========================================================================
  # Encapsulates business logic for creating organizations. In a real app,
  # this might handle slug generation, default settings, provisioning
  # external resources (Stripe customer, S3 bucket), audit logging, etc.

  alias AutonomaExample.Repo
  alias AutonomaExample.Schemas.Organization

  @doc """
  Create an organization with business logic applied.

  In production this would also:
    - Generate a URL-safe slug from the name
    - Create default settings / billing records
    - Provision external resources (Stripe, etc.)
  """
  def create(attrs) do
    %Organization{}
    |> Ecto.Changeset.cast(attrs, [:id, :name])
    |> Ecto.Changeset.validate_required([:name])
    |> Repo.insert!()
    |> to_map()
  end

  @doc """
  Delete an organization and clean up associated resources.

  In production this would also tear down external resources,
  cancel subscriptions, etc.
  """
  def delete(id) do
    Repo.get!(Organization, id)
    |> Repo.delete!()

    :ok
  end

  defp to_map(%Organization{} = org) do
    %{
      "id" => org.id,
      "name" => org.name
    }
  end
end

File: examples/typescript/nextjs-drizzle/src/repositories/organization.ts
// =============================================================================
// Organization Repository (Drizzle)
// =============================================================================
// A typical repository that wraps Drizzle with business logic.
// In a real app, this might generate slugs, set up billing, create default
// settings, or call external services (e.g., Stripe customer creation).

import { db } from '@/db'
import { organizations } from '@/db/schema'
import { eq } from 'drizzle-orm'

export class OrganizationRepository {
  async create(data: { name: string }) {
    const [organization] = await db
      .insert(organizations)
      .values({
        name: data.name,
        // In a real app you might also:
        // - Create a Stripe customer
        // - Set up default organization settings
        // - Send a welcome email to the creator
      })
      .returning()

    return organization
  }

  async delete(id: string) {
    // Business logic: clean up external resources before deleting
    // In a real app: cancel Stripe subscription, revoke API keys, etc.
    const [deleted] = await db
      .delete(organizations)
      .where(eq(organizations.id, id))
      .returning()

    return deleted
  }
}
